### PR TITLE
Design and implementation of the core credentials plugin API.

### PR DESCRIPTION
**Pull request by jboeuf**
- grpc_security.h defines some new data structures to plugin user defined client side metadata only credentials. 
- The goal is to have the wrapped languages use this API in order to allow for native (as in wrapped languages) credentials to be plugged-in. For example, a ruby user would write a function in ruby to populate auth metadata.
- We use C++ as an example to show how this API can be used while still
  providing an idiomatic interface in the wrapped language of choice.

### CODE REVIEW

File: include/grpc/grpc_security.h [hunk 1]
```diff
@@ -131,6 +131,32 @@ grpc_credentials *grpc_google_iam_credentials_create(
...
+typedef void (*grpc_credentials_plugin_metadata_cb)(
+    void *user_data, const grpc_metadata *creds_md, size_t num_creds_md,
+    grpc_status_code status, const char *error_details);
+
+typedef struct {
+  /* The implementation of this method has to be non-blocking.
+     - service_url is the fully qualified URL that the client stack is
+       connecting to.
+     - cb is the callback that needs to be called when the metadata is ready.
+     - user_data needs to be passed as the first parameter of the callback. */
```
Comment (jtattermusch) [inline include/grpc/grpc_security.h]: the "user_data" is basically a context object. Given it's unclear who's the "user" here (C core being "user" is a bit confusing), wouldn't it be better to call the param just "context" or "opaque"?

Comment (jboeuf) [inline include/grpc/grpc_security.h]: The naming is consistent with the rest of the code. We *could* change it to `context` but I think that would be quite a big renaming change throughout the code.

File: include/grpc/grpc_security.h [hunk 2]
```diff
@@ -131,6 +131,32 @@ grpc_credentials *grpc_google_iam_credentials_create(
     const char *authorization_token, const char *authority_selector,
     void *reserved);
 
+/* Callback function to be called by the metadata credentials plugin
+   implementation when the metadata is ready. */
+typedef void (*grpc_credentials_plugin_metadata_cb)(
+    void *user_data, const grpc_metadata *creds_md, size_t num_creds_md,
+    grpc_status_code status, const char *error_details);
```
Comment (jtattermusch) [inline include/grpc/grpc_security.h]: please document what is the status and error_details for.

File: include/grpc/grpc_security.h (hunk 1 above)
Comment (jboeuf) [inline include/grpc/grpc_security.h]: Added more documentation so that the use of these parameters is clearer.

**Comment by jboeuf** [comment]
@jtattermusch, @tbetbetbe. @nathanielmanistaatgoogle, @murgatroid99, @ctiller may want to have a look as well.

**Comment by jboeuf** [comment]
Zookeeper tests are failing. I certainly didn't touch that...
```
[ RUN      ] ZookeeperTest.ZookeeperStateChangeTwoRpc
D0915 23:25:25.010103152   12727 zookeeper_test.cc:104]      localhost:2181
2015-09-15 23:25:25,041:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [127.0.0.1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
2015-09-15 23:25:25,041:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [::1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
E0915 23:25:25.053444886   12727 tcp_client_posix.c:159]     socket error: connection refused
E0915 23:25:25.181529984   12727 tcp_client_posix.c:159]     socket error: connection refused
2015-09-15 23:25:25,201:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [127.0.0.1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
2015-09-15 23:25:25,202:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [::1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
2015-09-15 23:25:25,202:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [127.0.0.1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
2015-09-15 23:25:25,202:12727(0x7ff0f3cb3700):ZOO_ERROR@handle_socket_error_msg@1697: Socket [::1:2181] zk retcode=-4, errno=111(Connection refused): server refused to accept the client
E0915 23:25:25.202424217   12727 zookeeper_test.cc:139]      assertion failed: status == 0

FAILED: bins/dbg/zookeeper_test [ret=-6, pid=12727]
```

**Comment by jtattermusch** [comment]
Zookeeper might be failing because of this: https://github.com/grpc/grpc/pull/3344

File: src/cpp/client/secure_credentials.cc [hunk 3]
```diff
@@ -144,4 +144,64 @@ std::shared_ptr<Credentials> CompositeCredentials(
...
+  Status status = plugin_->GetMetadata(service_url, &metadata);
+  std::vector<grpc_metadata> md;
+  for (auto it = metadata.begin(); it != metadata.end(); ++it) {
+    md.push_back({it->first.c_str(),
+                  it->second.data(),
+                  it->second.size(),
+                  0,
+                  {{nullptr, nullptr, nullptr, nullptr}}});
+  }
+  cb(user_data, &md[0], md.size(),
```
Comment (yang-g) [inline src/cpp/client/secure_credentials.cc]: If md ends up being empty, some compiler will complain about md[0]. I think md should not be empty usually but better protect against it?

File: include/grpc++/security/credentials.h [hunk 4]
```diff
@@ -165,6 +168,24 @@ std::shared_ptr<Credentials> CompositeCredentials(
...
+// User defined metadata credentials.
+class MetadataCredentialsPlugin {
+ public:
+  virtual ~MetadataCredentialsPlugin() {}
+
+  // If this method returns true, the Process function will be scheduled in
+  // a different thread from the one processing the call.
+  virtual bool IsBlocking() const { return true; }
+
+  // Gets the auth metatada produced by this plugin. */
```
Comment (yang-g) [inline include/grpc++/security/credentials.h]: Should it say "Sets the auth metadata ..." instead? The ```*/``` at the end of line is redundant.

**Comment by yang-g** [comment]
If I have N plugins and then N credentials and composite then apply to an rpc, everything should just work right?

**Comment by yang-g** [comment]
LGTM

File: src/cpp/client/secure_credentials.cc (hunk 3 above)
Comment (jboeuf) [inline src/cpp/client/secure_credentials.cc]: good one. fixed.

File: include/grpc++/security/credentials.h (hunk 4 above)
Comment (jboeuf) [inline include/grpc++/security/credentials.h]: Removed the `*/`, thanks!

I don't think that it should be "Sets the auth metadata...". We're getting the metadata here by giving the address of a map that will be populated by the plugin.

**Comment by jboeuf** [comment]
@yang-g 
> If I have N plugins and then N credentials and composite then apply to an rpc, everything should just work right?
Yes definitely. These credentials are composable with any other credentials including a credentials of the same class.

**Comment by jtattermusch** [comment]
will this get merged soon?

**Pull request merged**